Measure performance of ESPResSo LB GPU vs. waLBerla LB CPU

Measure the performance for a fluid and a fluid with particle coupling, using the ESPResSo LB implementation on GPU single-precision and the walberla implementation on CPU double-precision (and single-precision if possible), on the same workstation, e.g. `lama`.

@reinaual @RudolfWeeber
The waLBerla branch was ported from `double` to `real_t` in [jngrad/espresso:walberla-real_t](https://github.com/jngrad/espresso/compare/1fba07b91e...walberla-real_t). To build with single precision, do:
```sh
cmake .. -DWITH_CCACHE=ON -DWITH_WALBERLA=ON -DWALBERLA_DOUBLE_ACCURACY=OFF -DWITH_STOKESIAN_DYNAMICS=OFF
```

Several LB tests fail due to rounding errors, with absolute differences of the order 10^-5 to 10^-8.

<details>
  <summary>CTest output: one unit test and 6 python tests failed</summary>

```
$> make check_unit_tests
85/85 Test #65: LbWalberla_test ..................   ***Failed   27.46 sec
/work/jgrad/espresso-fork-walberla/src/core/unit_tests/LbWalberla.cpp(455): error:
  in "forces_book_keeping/_1": absolute value of
  (*(lb->get_node_last_applied_force(cn, true)) - f).norm(){4.097141941644287e-08} exceeds 1e-10
Failure occurred in a following context:
    lb_generator = ; 

*** 220293 failures are detected in the test module "Walberla node setters and getters test"

$> make check_python
  1/141 Test  #29: engine_lb ...........................................***Failed    3.44 sec
======================================================================
FAIL: test_momentum_conservation (__main__.SwimmerTestWALBERLA)
----------------------------------------------------------------------
Mismatched elements: 3 / 3 (100%)
Max absolute difference: 5.3081567e-05
Max relative difference: inf
 x: array([-5.308157e-05,  1.389081e-05,  2.329607e-05])
 y: array([0., 0., 0.])

  2/141 Test  #68: lb ..................................................***Failed    2.77 sec
======================================================================
FAIL: test_ext_force_density (__main__.TestLBWalberla)
----------------------------------------------------------------------
Mismatched elements: 3 / 3 (100%)
Max absolute difference: 8.2433641e-05
Max relative difference: 0.00019077
 x: array([7.451918, 3.887921, 0.324062])
 y: array([7.452, 3.888, 0.324])

======================================================================
FAIL: test_thermalization_force_balance (__main__.TestLBWalberla)
----------------------------------------------------------------------
Mismatched elements: 1 / 3 (33.3%)
Max absolute difference: 4.94645947e-06
Max relative difference: 1.07000398e-07
 x: array([-15.432864, 181.855323, 125.354238])
 y: array([-15.432866, 181.855324, 125.354233])

 11/141 Test  #69: lb_walberla .........................................***Failed    3.80 sec
======================================================================
FAIL: test (__main__.LbWalberlaTest)
----------------------------------------------------------------------
AssertionError: 1.3000001549720763 != 1.3 within 1e-10 delta

 12/141 Test #119: lb_vtk ..............................................***Failed    5.37 sec
======================================================================
FAIL: test_vtk (__main__.TestVTK)
----------------------------------------------------------------------
Mismatched elements: 2 / 12 (16.7%)
Max absolute difference: 1.4901161e-08
Max relative difference: 2.4836081e-05
 x: array([0.0006 , 0.00075, 0.00075, 0.00075, 0.00075, 0.00075, 0.00075,
       0.00075, 0.00075, 0.00075, 0.00075, 0.0006 ], dtype=float32)
 y: array([0.0006 , 0.00075, 0.00075, 0.00075, 0.00075, 0.00075, 0.00075,
       0.00075, 0.00075, 0.00075, 0.00075, 0.0006 ], dtype=float32)

120/141 Test  #95: observable_profileLB ................................***Failed    4.23 sec
======================================================================
FAIL: test_velocity_profile (__main__.LBWalberla)
----------------------------------------------------------------------
Mismatched elements: 1 / 3 (33.3%)
Max absolute difference: 2.74181366e-06
Max relative difference: 2.49255787e-07
 x: array([1.100000e+01, 2.086162e-07, 0.000000e+00])
 y: array([11.,  0.,  0.])

121/141 Test #116: lb_shear ............................................***Failed    3.43 sec
======================================================================
FAIL: test_xy (__main__.LBWalberlaShear)
----------------------------------------------------------------------
Mismatched elements: 1 / 3 (33.3%)
Max absolute difference: 0.00029982
Max relative difference: 3.
 x: array([-8.942400e+03,  6.458365e+00,  1.998782e-04])
 y: array([-8.942400e+03,  6.458465e+00, -9.993911e-05])

======================================================================
FAIL: test_xz (__main__.LBWalberlaShear)
----------------------------------------------------------------------
Mismatched elements: 1 / 3 (33.3%)
Max absolute difference: 0.00070513
Max relative difference: 1.9203658e+09
 x: array([-8.942400e+03,  9.993911e-05,  6.458565e+00])
 y: array([-8.942400e+03, -5.204170e-14,  6.458365e+00])

======================================================================
FAIL: test_yz (__main__.LBWalberlaShear)
----------------------------------------------------------------------
Mismatched elements: 2 / 3 (66.7%)
Max absolute difference: 0.0007662
Max relative difference: 1.
 x: array([ 4.163336e-14, -8.942400e+03, -6.458265e+00])
 y: array([ 1.998782e-04, -8.942401e+03, -6.458465e+00])
```
</details>

--- a/maintainer/benchmarks/CMakeLists.txt
+++ b/maintainer/benchmarks/CMakeLists.txt
@@ -96,6 +96,10 @@ python_benchmark(
   FILE p3m.py ARGUMENTS
   "--particles_per_core=10000;--volume_fraction=0.25;--prefactor=4")
 
+python_benchmark(
+  FILE lb.py ARGUMENTS
+  "--particles_per_core=125;--volume_fraction=0.03;--lb_sites_per_core=28")
+
 add_custom_target(
   benchmark_python COMMAND ${CMAKE_CTEST_COMMAND} --timeout ${TEST_TIMEOUT}
                            ${CTEST_ARGS} --output-on-failure)

--- a/maintainer/benchmarks/lb.py
+++ b/maintainer/benchmarks/lb.py
@@ -1,0 +1,183 @@
+#
+# Copyright (C) 2013-2019 The ESPResSo project
+#
+# This file is part of ESPResSo.
+#
+# ESPResSo is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# ESPResSo is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+#
+
+"""
+Benchmark Lattice-Boltzmann fluid + Lennard-Jones particles
+"""
+import os
+import sys
+import numpy as np
+from time import time
+import argparse
+
+parser = argparse.ArgumentParser(description="Benchmark LB simulations. "
+                                 "Save the results to a CSV file.")
+parser.add_argument("--particles_per_core", metavar="N", action="store",
+                    type=int, default=125, required=False,
+                    help="Number of particles in the simulation box")
+parser.add_argument("--lb_sites_per_particle", metavar="N_LB", action="store",
+                    type=float, default=28, required=False,
+                    help="Number of particles in the simulation box")
+parser.add_argument("--volume_fraction", metavar="FRAC", action="store",
+                    type=float, default=0.03, required=False,
+                    help="Fraction of the simulation box volume occupied by "
+                    "particles (range: [0.01-0.74], default: 0.50)")
+group = parser.add_mutually_exclusive_group()
+group.add_argument("--output", metavar="FILEPATH", action="store",
+                   type=str, required=False, default="benchmarks.csv",
+                   help="Output file (default: benchmarks.csv)")
+
+args = parser.parse_args()
+
+# process and check arguments
+n_iterations = 30
+assert args.volume_fraction > 0, "volume_fraction must be a positive number"
+assert args.volume_fraction < np.pi / (3 * np.sqrt(2)), \
+    "volume_fraction exceeds the physical limit of sphere packing (~0.74)"
+
+
+import espressomd
+required_features = ["LENNARD_JONES"]
+espressomd.assert_features(required_features)
+
+
+# System
+#############################################################
+system = espressomd.System(box_l=[1, 1, 1])
+
+# Interaction parameters (Lennard-Jones)
+#############################################################
+
+lj_eps = 1.0  # LJ epsilon
+lj_sig = 1.0  # particle diameter
+lj_cut = lj_sig * 2**(1. / 6.)  # cutoff distance
+
+# System parameters
+#############################################################
+
+n_proc = system.cell_system.get_state()['n_nodes']
+n_part = n_proc * args.particles_per_core
+# volume of N spheres with radius r: N * (4/3*pi*r^3)
+box_l = (n_part * 4. / 3. * np.pi * (lj_sig / 2.)**3
+         / args.volume_fraction)**(1. / 3.)
+lb_grid = int((round(n_part * args.lb_sites_per_particle)**(1. / 3)))
+agrid = box_l / lb_grid
+measurement_steps = int(max(120**3 / lb_grid**3, 50))
+
+# System
+#############################################################
+system.box_l = 3 * (box_l,)
+
+# PRNG seeds
+#############################################################
+# np.random.seed(1)
+
+# Integration parameters
+#############################################################
+system.time_step = 0.01
+system.cell_system.skin = 0.5
+system.thermostat.turn_off()
+
+
+#############################################################
+#  Setup System                                             #
+#############################################################
+
+# Interaction setup
+#############################################################
+system.non_bonded_inter[0, 0].lennard_jones.set_params(
+    epsilon=lj_eps, sigma=lj_sig, cutoff=lj_cut, shift="auto")
+
+
+# Particle setup
+#############################################################
+
+#  Warmup Integration                                       #
+#############################################################
+
+system.integrator.set_steepest_descent(
+    f_max=0,
+    gamma=0.001,
+    max_displacement=0.01)
+
+# warmup
+while system.analysis.energy()["total"] > 0.1 * n_part:
+    print("minimization: {:.1f}".format(system.analysis.energy()["total"]))
+    system.integrator.run(20)
+print("minimization: {:.1f}".format(system.analysis.energy()["total"]))
+print()
+system.integrator.set_vv()
+
+system.thermostat.set_langevin(kT=1.0, gamma=1.0, seed=42)
+
+# tuning and equilibration
+print("Tune skin: {}".format(system.cell_system.tune_skin(
+    min_skin=0.2, max_skin=1, tol=0.05, int_steps=100)))
+system.integrator.run(500)
+print("Tune skin: {}".format(system.cell_system.tune_skin(
+    min_skin=0.2, max_skin=1, tol=0.05, int_steps=100)))
+system.integrator.run(500)
+
+
+system.thermostat.turn_off()
+print("lb sites", lb_grid, "agrid", agrid)
+if "LBFluid" in dir(espressomd.lb):
+    LBClass = espressomd.lb.LBFluid
+elif "LBFluidWalberla" in dir(espressomd.lb):
+    LBClass = espressomd.lb.LBFluidWalberla
+else: 
+    raise Exception("LB not built in")
+
+lbf = LBClass(agrid=agrid, dens=1, visc=1, tau=system.time_step, kT=1, seed=1)
+system.actors.add(lbf)
+print("lb shape", lbf.shape)
+system.thermostat.set_lb(gamma=10, LB_fluid=lbf, seed=2)
+
+
+# time integration loop
+print("Timing every {} steps".format(measurement_steps))
+main_tick = time()
+all_t = []
+for i in range(n_iterations):
+    tick = time()
+    system.integrator.run(measurement_steps)
+    tock = time()
+    t = (tock - tick) / measurement_steps
+    print("step {}, time = {:.2e}, verlet: {:.2f}, energy: {:.2e}"
+          .format(i, t, system.cell_system.get_state()["verlet_reuse"],
+                  system.analysis.energy()["total"]))
+    all_t.append(t)
+main_tock = time()
+# average time
+all_t = np.array(all_t)
+avg = np.average(all_t)
+ci = 1.96 * np.std(all_t) / np.sqrt(len(all_t) - 1)
+print("average: {:.3e} +/- {:.3e} (95% C.I.)".format(avg, ci))
+
+cmd = " ".join(x for x in sys.argv[1:] if not x.startswith("--output"))
+report = ('"{script}","{arguments}",{cores},{mean:.3e},'
+          '{ci:.3e},{n},{dur:.1f}\n'.format(
+              script=os.path.basename(sys.argv[0]), arguments=cmd,
+              cores=n_proc, dur=main_tock - main_tick, n=measurement_steps,
+              mean=avg, ci=ci))
+if not os.path.isfile(args.output):
+    report = ('"script","arguments","cores","mean","ci",'
+              '"nsteps","duration"\n' + report)
+with open(args.output, "a") as f:
+    f.write(report)

